File: src/plnflr/engine/plank.py
"""Lay planks: bbox grid clipped to the inset polygon."""

from __future__ import annotations

from dataclasses import replace
from typing import Literal

from plnflr.domain.models import (
    LayoutPlan,
    LayoutRules,
    Piece,
    PlankSpec,
    Room,
    Vertex,
    Warning,
)
from plnflr.engine.bom import make_bom
from plnflr.engine.clip import area_mm2, intersect_rect, ring_to_tuple
from plnflr.engine.expansion import bbox, inset_room, resolve_gap_mm
from plnflr.engine.grid import iter_slots
from plnflr.engine.instructions import instructions_for
from plnflr.engine.rotate import origin_of, rotate_points, rotate_room


def _direction(room: Room, rules: LayoutRules) -> Literal["along_x", "along_y"]:
    if rules.direction in {"along_x", "along_y"}:
        return rules.direction
    min_x, min_y, max_x, max_y = bbox(room)
    along_long = (max_x - min_x) >= (max_y - min_y)
    if rules.direction == "along_short":
        along_long = not along_long
    return "along_x" if along_long else "along_y"


def _lay_axis(
    inner: Room, spec: PlankSpec, rules: LayoutRules
) -> tuple[tuple[Piece, ...], Literal["along_x", "along_y"], list[Warning]]:
    direction = _direction(inner, rules)
    min_x, min_y, max_x, max_y = bbox(inner)
    span = max(max_x - min_x, max_y - min_y)
    warnings: list[Warning] = []
    if span > rules.intermediate_joint_mm:
        metres = f"{span / 1000:.3f}".replace(".", ",")
        warnings.append(
            Warning(
                "intermediate_joint",
                f"Pomieszczenie ma {metres} m w świetle po dylatacji. "
                "Rozważ przerwę dylatacyjną / profil co ok. 8 m.",
            )
        )
    slots = iter_slots(
        min_x=min_x,
        min_y=min_y,
        max_x=max_x,
        max_y=max_y,
        plank_length_mm=spec.length_mm,
        plank_width_mm=spec.width_mm,
        min_row_width_mm=rules.min_row_width_mm,
        min_end_mm=rules.min_end_length_mm,
        stagger=rules.stagger,
        direction=direction,
    )
    outer_path = ring_to_tuple(inner.outer)
    hole_paths = tuple(ring_to_tuple(h) for h in inner.holes)
    pieces: list[Piece] = []
    board = 0
    order = 0
    for slot in slots:
        rect = (
            (slot.x0, slot.y0),
            (slot.x1, slot.y0),
            (slot.x1, slot.y1),
            (slot.x0, slot.y1),
        )
        clipped = intersect_rect(rect, outer_path, hole_paths)
        if not clipped:
            continue
        board += 1
        for poly in clipped:
            order += 1
            kind = slot.kind if len(poly) == 4 else "clip"
            along = (slot.x1 - slot.x0) if direction == "along_x" else (slot.y1 - slot.y0)
            across = (slot.y1 - slot.y0) if direction == "along_x" else (slot.x1 - slot.x0)
            pieces.append(
                Piece(
                    piece_id=f"r{slot.row_index}-b{board}-{order}",
                    row_index=slot.row_index,
                    geometry=tuple(Vertex(x, y) for x, y in poly),
                    kind=kind,  # type: ignore[arg-type]
                    source_board=board,
                    install_order=order,
                    length_mm=along,
                    width_mm=across,
                )
            )
    return tuple(pieces), direction, warnings


def layout_planks(room: Room, spec: PlankSpec, rules: LayoutRules) -> LayoutPlan:
    gap = resolve_gap_mm(room, rules)
    inner = inset_room(room, gap)
    angle = int(rules.angle_deg) % 360
    origin = origin_of(inner)
    grid_room = rotate_room(inner, angle, origin) if angle else inner
    pieces, direction, warnings = _lay_axis(grid_room, spec, rules)
    if angle:
        pieces = tuple(
            replace(
                piece,
                geometry=tuple(
                    Vertex(x, y)
                    for x, y in rotate_points(
                        tuple((v.x_mm, v.y_mm) for v in piece.geometry),
                        -angle,
                        origin,
                    )
                ),
            )
            for piece in pieces
        )
    outer_path = ring_to_tuple(inner.outer)
    hole_paths = tuple(ring_to_tuple(h) for h in inner.holes)
    net = area_mm2(outer_path) - sum(area_mm2(h) for h in hole_paths)
    bom = make_bom(
        pieces=len(pieces),
        full_boards=max((p.source_board for p in pieces), default=0),
        boards_per_pack=spec.boards_per_pack,
        area_net_mm2=net,
        board_area_mm2=spec.length_mm * spec.width_mm,
        label="Panele",
        kind="plank",
    )
    min_x, min_y, max_x, max_y = bbox(inner)
    longer_x = (max_x - min_x) >= (max_y - min_y)
    axis = "dłuższego" if direction == "along_x" and longer_x else "wybranego"
    angle_note = f" Kąt {angle}°." if angle else ""
    rationale = (
        f"Kierunek {direction.replace('_', ' ')} — deski wzdłuż {axis} boku. "
        f"Dylatacja {gap} mm. Siatka na bbox, przycięcie do obrysu.{angle_note}"
    )
    return LayoutPlan(
        room=room,
        gap_mm=gap,
        inset=inner,
        direction=direction,
        pieces=pieces,
        bom=bom,
        warnings=tuple(warnings),
        rationale_pl=rationale,
        rows_instruction_pl=instructions_for(
            pieces, gap_mm=gap, has_holes=bool(room.holes)
        ),
        angle_deg=angle,
        boms=(bom,),
    )

File: Kernel/Sources/PlnFlrLayout/Plank.swift
func originOf(_ room: Room) -> Vertex {
    let bounds = bbox(room)
    return Vertex((bounds.minX + bounds.maxX) / 2, (bounds.minY + bounds.maxY) / 2)
}

func rotateRoom(_ room: Room, angleDeg: Int, origin: Vertex) throws -> Room {
    guard angleDeg % 360 != 0 else { return room }
    let outer = try Ring(rotatePoints(room.outer.vertices, angleDeg: angleDeg, origin: origin))
    let holes = try room.holes.map {
        try Ring(rotatePoints($0.vertices, angleDeg: angleDeg, origin: origin))
    }
    return Room(outer, holes: holes)
}

func direction(of room: Room, rules: LayoutRules, windows: [Opening]) -> Axis {
    if rules.direction == .alongX { return .alongX }
    if rules.direction == .alongY { return .alongY }
    if rules.direction == .intoWindow, let window = windows.max(by: {
        abs($0.end.xMm - $0.start.xMm) + abs($0.end.yMm - $0.start.yMm)
            < abs($1.end.xMm - $1.start.xMm) + abs($1.end.yMm - $1.start.yMm)
    }) {
        let alongWallX = abs(window.end.xMm - window.start.xMm)
            >= abs(window.end.yMm - window.start.yMm)
        return alongWallX ? .alongY : .alongX
    }
    let bounds = bbox(room)
    var alongLong = (bounds.maxX - bounds.minX) >= (bounds.maxY - bounds.minY)
    if rules.direction == .alongShort { alongLong.toggle() }
    return alongLong ? .alongX : .alongY
}

public func layoutPlanks(
    _ room: Room,
    _ spec: PlankSpec,
    _ rules: LayoutRules,
    windows: [Opening] = []
) throws -> LayoutPlan {
    let gap = try resolveGapMm(room, rules)
    let inner = try insetRoom(room, gapMm: gap)
    let angle = ((rules.angleDeg % 360) + 360) % 360
    let origin = originOf(inner)
    let gridRoom = angle == 0 ? inner : try rotateRoom(inner, angleDeg: angle, origin: origin)
    let axis = direction(of: gridRoom, rules: rules, windows: windows)
    let bounds = bbox(gridRoom)
    let span = max(bounds.maxX - bounds.minX, bounds.maxY - bounds.minY)
    var warnings: [Warning] = []
    if span > rules.intermediateJointMm {
        let metres = String(format: "%.3f", Double(span) / 1000).replacingOccurrences(
            of: ".",
            with: ","
        )
        warnings.append(
            Warning(
                code: "intermediate_joint",
                messagePl:
                    "Pomieszczenie ma \(metres) m w świetle po dylatacji. Rozważ przerwę dylatacyjną / profil co ok. 8 m."
            )
        )
    }
    let slots = try iterSlots(
        minX: bounds.minX,
        minY: bounds.minY,
        maxX: bounds.maxX,
        maxY: bounds.maxY,
        plankLengthMm: spec.lengthMm,
        plankWidthMm: spec.widthMm,
        minRowWidthMm: rules.minRowWidthMm,
        minEndMm: rules.minEndLengthMm,
        stagger: rules.stagger,
        direction: axis
    )
    let outerPath = gridRoom.outer.vertices
    let holePaths = gridRoom.holes.map(\.vertices)
    var pieces: [Piece] = []
    var board = 0
    var order = 0
    for slot in slots {
        let rect = [
            Vertex(slot.x0, slot.y0),
            Vertex(slot.x1, slot.y0),
            Vertex(slot.x1, slot.y1),
            Vertex(slot.x0, slot.y1),
        ]
        let clipped = try intersectRect(rect, outer: outerPath, holes: holePaths)
        guard !clipped.isEmpty else { continue }
        board += 1
        for poly in clipped {
            order += 1
            let kind: PieceKind = poly.count == 4 ? slot.kind : .clip
            let along = axis == .alongX ? slot.x1 - slot.x0 : slot.y1 - slot.y0
            let across = axis == .alongX ? slot.y1 - slot.y0 : slot.x1 - slot.x0
            var geometry = poly
            if angle != 0 {
                geometry = rotatePoints(poly, angleDeg: -angle, origin: origin)
            }
            pieces.append(
                Piece(
                    geometry: geometry,
                    installOrder: order,
                    kind: kind,
                    lengthMm: along,
                    pieceId: "r\(slot.rowIndex)-b\(board)-\(order)",
                    rowIndex: slot.rowIndex,
                    sourceBoard: board,
                    widthMm: across
                )
            )
        }
    }
    let net = areaMm2(inner.outer.vertices)
        - inner.holes.reduce(0) { $0 + areaMm2($1.vertices) }
    let bom = makeBom(
        pieces: pieces.count,
        fullBoards: pieces.map(\.sourceBoard).max() ?? 0,
        boardsPerPack: spec.boardsPerPack,
        areaNetMm2: net,
        boardAreaMm2: spec.lengthMm * spec.widthMm
    )
    let innerBounds = bbox(inner)
    let longerX = (innerBounds.maxX - innerBounds.minX) >= (innerBounds.maxY - innerBounds.minY)
    let axisWord: String
    if rules.direction == .intoWindow, !windows.isEmpty {
        axisWord = "prostopadle do ściany z oknem"
    } else {
        axisWord = axis == .alongX && longerX ? "dłuższego" : "wybranego"
    }
    let angleNote = angle == 0 ? "" : " Kąt \(angle)°."
    let rationale: String
    if rules.direction == .intoWindow, !windows.isEmpty {
        rationale =
            "Deski prostopadle do okna (\(axisWord)). Dylatacja \(gap) mm. Siatka na bbox, przycięcie do obrysu.\(angleNote)"
    } else {
        rationale =
            "Kierunek \(axis.rawValue.replacingOccurrences(of: "_", with: " ")) — deski wzdłuż \(axisWord) boku. Dylatacja \(gap) mm. Siatka na bbox, przycięcie do obrysu.\(angleNote)"
    }
    return LayoutPlan(
        bom: bom,
        direction: axis,
        gapMm: gap,
        inset: inner,
        pieces: pieces,
        rationalePl: rationale,
        room: room,
        rowsInstructionPl: instructionsFor(pieces, gapMm: gap, hasHoles: !room.holes.isEmpty),
        warnings: warnings,
        windows: windows
    )
}
